File: java/src/main/java/kata-bowling-game/Game.java

package kata_bowling_game;

public class Game {

    public static final int MAX_PINS_IN_FRAME = 10;
    private static final int MAX_FLAME = MAX_PINS_IN_FRAME;
    private int rolls[] = new int[21];
    private int currentRoll = 0;

    Game() {
    }

    public void roll(int pins) {
        rolls[currentRoll++] = pins;
    }

    public int score() {
        int score = 0;
        int rollIndex = 0;
        for (int frame = 0; frame < MAX_FLAME; frame++) {
            if (isStrike(rollIndex)) {
                score += MAX_PINS_IN_FRAME + strikeBonus(rollIndex);
                rollIndex += 1;
            } else if (isSpare(rollIndex)) {
                score += MAX_PINS_IN_FRAME + spareBonus(rollIndex);
                rollIndex += 2;
            } else {
                score += sumOfPinsInFrame(rollIndex);
                rollIndex += 2;
            }
        }
        return score;
    }

    private int spareBonus(int rollIndex) {
        return rolls[rollIndex + 2];
    }

    private int strikeBonus(int rollIndex) {
        return rolls[rollIndex + 1] + rolls[rollIndex + 2];
    }

    private int sumOfPinsInFrame(int rollIndex) {
        return rolls[rollIndex] + rolls[rollIndex+1];
    }

    private boolean isStrike(int rollIndex) {
        return rolls[rollIndex] == MAX_PINS_IN_FRAME;
    }

    private boolean isSpare(int rollIndex) {
        return sumOfPinsInFrame(rollIndex) == MAX_PINS_IN_FRAME;
    }

}

File: bowling-game/scala/src/main/scala/example/Game.scala
package example

object Game {

  def score(rolls: List[Int]): Int = rolls match {
    case r1 :: r2 :: r3 :: tail if isStrike(r1) => 10 + r2 + r3 + score(r2 :: r3 :: tail)
    case r1 :: r2 :: r3 :: tail if isSpare(r1, r2) => 10 + r3 + score(r3 :: tail)
    case r :: tail => r + score(tail)
    case _ => 0
  }

  private def isStrike(roll: Int) = roll == 10


  private def isSpare(roll1: Int, roll2: Int) = roll1 + roll2 == 10

}
